---
title: "Calculating genotype likelihoods and posteriors from read depths and base quality scores"
output: html_document
runtime: shiny
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(cowplot)
dump <- lapply(dir("R", pattern = "*.R", full.names = TRUE), function(f) source(f))
```

In the brave new world of next generation sequencing, we may not be able to
observe the genotype fully and accurately.  Instead, from each individual we
get some number (possibly 0) of reads that cover the site we are interested
in.  If a lot of those reads have a $T$ at the site, and few of them have a $C$
at the site, it is likely the individual is $TT$.  If roughly the same number
of reads have $C$'s and $T$'s it is likely the individual is $CT$, and so on.
However, it is a little more complex, because we naturally would want to
give higher weight to reads with higher quality scores. Making use of this
read information requires a model by which we can compute the _likelihood_ of
the individual's genotype given the read data.

This notebook has some simple exercises to familiarize the reader with
how genotype likelihoods (and genotype posteriors) behave.

We adopt the SAMtools genotyping error model for two possible alleles.
Most of the other genotyping error models are similar.

## Assume prior parameters, $\alpha_T$ and $\alpha_C$

We start, as before, with a prior for allele frequency.
```{r, echo=FALSE}
inputPanel(
  numericInput(
    "alpha_1", 
    label = "alpha_T",
    min = 0.00, 
    max = 100, 
    value = 0.5,
    step = 0.1
  ),
  numericInput(
    "alpha_2", 
    label = "alpha_C",
    min = 0.00, 
    max = 100, 
    value = 0.5,
    step = 0.1
  )
)
```

### Here is what the prior distribution looks like:

```{r, echo=FALSE, out.height=100}
renderPlot({
  ggplot() +
    xlim(0, 1) +
    geom_function(
      fun = dbeta,
      args = list(
        shape1 = input$alpha_1,
        shape2 = input$alpha_2
      ),
      n = 5000
    )
  }, 
  height = 150 
)
```



## Set a true allele frequency, simulate some genotypes, simulate reads from those genotypes, and witness the genotype likelihoods

Just like in `001-allele-freq-estimation` we are going to simulate
a sample of genotypes.  But here, those genotypes will not be observed---they
are _latent variables_. What we get to observe are the read depths, and we
will simulate those to have such data to play with.

Set $p_T$ and $n$, as before, and simulate the latent genotypes by
pressing `Simulate new genotypes`. 

Then, hit
`Simulate reads` to simulate the observed read information. You can set the
average read depth per individual, and you change how much read depth
varies between individuals. `R.D. "variability" across indivs` = 0 means not much variation in
read depth between individuals; 1 means quite a lot of
variation in read depth between individuals.

```{r, echo=FALSE}
inputPanel(
  numericInput(
    "p_T", 
    label = "p_T",
    min = 0.00, 
    max = 1.0, 
    value = 0.3,
    step = 0.01
  ),
  numericInput(
    "n", 
    label = "n",
    min = 1, 
    max = 200, 
    value = 60,
    step = 1
  ),
  actionButton(
    "sim_new", 
    "Simulate new genotypes",
    class = "btn-warning"
  )
)

inputPanel(
  numericInput(
    "ave_rd",
    label ="Average per-indiv read depth",
    min = 0.0,
    max = 50,
    value = 4,
    step = 1
  ),
  numericInput(
    "disp",
    label = "R.D. \"variability\" across indivs",
    min = 0,
    max = 1,
    value = 0.0,
    step = 0.05
  ),
  sliderInput(
    "phredQ_range", 
    label = "Range of PHRED quality scores",
    min = 0, 
    max = 40,
    value = c(20,40)),
  sliderInput(
    "phredQ_depart", 
    label = "Range of PHRED reality departure",
    min = -20, 
    max = 10,
    value = c(-1,1)),
  actionButton(
    "sim_new_rd", 
    "Simulate new reads",
    class = "btn-warning"
  )
)


# simulate new dataset of genotypes
genotypes <- eventReactive(
  input$sim_new,
  {
    max_in_row <- 20
    genos <- tibble(
      g012 = sample(
        x = 2:0,
        size = input$n,
        prob = c(
          input$p_T^2,
          2 * input$p_T * (1 - input$p_T), 
          (1 - input$p_T) ^ 2
        ),
        replace = TRUE
      )
    ) %>%
      arrange(g012) %>%
      mutate(idx = (1:n()) - 1) %>%
      mutate(
        geno = case_when(
           g012== 2 ~ "TT",
          g012 == 1 ~ "CT",
          g012 == 0 ~ "CC"
        ),
        ypos = -floor(idx / max_in_row),
        xpos = idx %% max_in_row
      )
    genos
      
  }
)

# simulate a new data set of read depths
reads_and_quals <- eventReactive(
  input$sim_new_rd,
  {
    genos <- genotypes() %>%
      ungroup()
    reads_and_quals <- genos %>%
      mutate(
        tot_rd = rnbinom(
          n(), 
          mu = input$ave_rd,
          size = 1/input$disp
        ),
        alle_list = str_split(geno, ""),
        sampled_allele = map2(
          .x = alle_list,
          .y = tot_rd,
          .f = function(x, y) sample(x, size = y, replace = TRUE)
        )
      ) %>%
      select(-alle_list) %>%
      unnest(sampled_allele) %>%
      mutate(
        phredQ = sample(
          x = input$phredQ_range[1]:input$phredQ_range[2], 
          size = n(),
          replace = TRUE
        ),
        phredQreal = phredQ + sample(
          x = input$phredQ_depart[1]:input$phredQ_depart[2], 
          size = n(),
          replace = TRUE
        ),
        phredQreal = ifelse(phredQreal < 0, 0, phredQreal),  # make sure none are less than 0
        seq_err_prob = 10 ^ -(phredQ/10),
        true_err_prob = 10 ^ -(phredQreal/10),
        sequenced_allele = case_when(
          runif(n()) > true_err_prob ~ sampled_allele,
          sampled_allele == "C" ~ "T",
          sampled_allele == "T" ~ "C"
        )
      )
    reads_and_quals
  }
)

renderPlot(
  {
    
    # get a summary of the read depths if they are available
    rq <- reads_and_quals() 
    genos <- genotypes()
    genolikes <- geno_likelihoods(rq, genos)
    
    
    pwl <- plot_with_likelihoods(input, genos, rq, genolikes)
    
    cowplot::plot_grid(
      pwl$geno_plot,
      pwl$post_freq_plot,
      nrow = 1,
      rel_widths = c(0.65, 0.35)
    )
  },
  height = 600
)
```

### Table of reads and quality scores

```{r, echo=FALSE}
 DT::renderDataTable(DT::datatable({
    rq <- reads_and_quals()
    
    rq %>%
      rename(
        true_geno = geno,
        tot_read_depth = tot_rd,
        seqed_allele = sequenced_allele,
        phredQ_seq = phredQ,
        phredQ_true = phredQreal
      ) %>%
      mutate(
        indiv = idx + 1,
        seq_err_prob = sprintf("%.7f", seq_err_prob),
        true_err_prob = sprintf("%.7f", true_err_prob)
      ) %>%
      group_by(idx) %>%
      mutate(read_num = 1:n()) %>%
      ungroup() %>%
      select(
        indiv,
        true_geno,
        #tot_read_depth,
        read_num,
        sampled_allele,
        seqed_allele,
        phredQ_seq,
        phredQ_true,
        seq_err_prob,
        true_err_prob
      )
  }))
```


## Compute and display the posterior probability of each genotype, report the maximum _a-posteriori_ genotype, and plot the sample from the posterior distribution of the allele frequencies.

```{r, echo=FALSE}
inputPanel(
 sliderInput(
    "burn_sweeps", 
    label = "burn_in (lo-value) and total sweeps (hi-value)",
    min = 0, 
    max = 5000,
    step = 100,
    value = c(100,1200)),
  actionButton(
    "calc_posts", 
    "Calculate Posteriors with MCMC",
    class = "btn-warning"
  )
)

# compute posteriors
posterior_sample <- eventReactive(
  input$calc_posts,
  {
    # get a summary of the read depths if they are available
    rq <- reads_and_quals() 
    genos <- genotypes()
    genolikes <- geno_likelihoods(rq, genos)
    
    compute_posteriors_mcmc(
      input, 
      genos, 
      rq, 
      genolikes
    )
    
  }
)

renderPlot(
  {
    
    # get a summary of the read depths if they are available
    rq <- reads_and_quals() 
    genos <- genotypes()
    genolikes <- geno_likelihoods(rq, genos)
    post_sample <- posterior_sample()
    
    
    pwl <- plot_with_likelihoods(input, genos, rq, genolikes)
    
    
    # save the outputs for future development, etc,
    genolikes2 <- pwl$genolikes2
    
    # add geno posteriors to the genolikes2
    genolikes3 <- genolikes2 %>%
      left_join(post_sample$GP, by = c("idx", "hyp_geno"))
    
    # make a data frame that has the max likelihood and max posterior
    # genotype calls.
    max_like_genos_count <- genolikes3 %>%
      group_by(idx) %>%
      mutate(
        no_data = ifelse(all(near(scaled_likelihood, 1/3)), TRUE, FALSE)
      ) %>%
      arrange(idx, desc(scaled_likelihood)) %>%
      slice(1) %>%
      ungroup() %>%
      mutate(inferred_genotype = ifelse(no_data, NA, hyp_geno)) %>%
      count(geno, inferred_genotype) %>%
      mutate(true_genotype = factor(geno, levels = c("TT", "CT", "CC")))
    
    max_post_genos <- genolikes3 %>%
      group_by(idx) %>%
      arrange(idx, desc(geno_posterior)) %>%
      slice(1) %>%
      ungroup() %>%
      mutate(inferred_genotype = hyp_geno) %>%
      select(idx, geno, inferred_genotype)
    
    max_post_genos_count <- max_post_genos %>%
      count(geno, inferred_genotype) %>%
      mutate(true_genotype = factor(geno, levels = c("TT", "CT", "CC")))
    
    genolikes4 <- genolikes3 %>%
      left_join(max_post_genos %>% select(idx, inferred_genotype), by = "idx")
      
    
    max_like_genos_plot <- ggplot(max_like_genos_count, aes(x = inferred_genotype, y = true_genotype)) +
      geom_text(aes(label = n)) +
      theme_bw() +
      ggtitle("Maximum likelihood genotypes (summary):")
    
    max_post_genos_plot <- ggplot(max_post_genos_count, aes(x = inferred_genotype, y = true_genotype)) +
      geom_text(aes(label = n)) +
      theme_bw() +
      ggtitle("Maximum a-posteriori genotypes (summary):")
             
    
    # save(
    #   genos,
    #   rq,
    #   genolikes, 
    #   post_sample,
    #   pwl,
    #   genolikes2,
    #   file = "test_data/vars.rda"
    # )
    
   
    
    # get the posterior frequency of the allele plot made
    post_freq_plot <- pwl$post_freq_plot +
      geom_histogram(
        data = post_sample$p,
        mapping = aes(
          x = p,
          y = ..density..
          ),
        fill = "blue",
        alpha = 0.3,
        binwidth = 0.01
      ) +
      ylab("Posterior density / MCMC histogram") + 
      ggtitle("Allele frequency posterior:")

    # make the top-row trace plot
    trace <- ggplot(
      post_sample$p,
      aes(x = sweep, y = p)
    ) + 
      geom_line() +
      ggtitle("Trace plot of frequency of T allele:")
    

    
    # now, add the posteriors to the geno plot. 
    geno_plot <- pwl$geno_plot +
      geom_rect(   # get the "bar plots" of the likelihoods
      data = genolikes4,
      aes(
        xmin = xl,
        xmax = xl + w,
        ymin = ybot,
        ymax = ybot + (geno_posterior * full_h),
      ),
      fill = NA,
      colour = "gray",
      size = 1
    ) +
      geom_label(
        data = genolikes4,
        aes(
          x = xpos,
          y = ybot + full_h,
          label = inferred_genotype,
          fill = as.character(inferred_genotype),
        ),
        colour = "white",
        fontface = "bold"
      )
    
    # start mashing up two different columns
    left_side <- plot_grid(
      trace, 
      geno_plot,
      ncol = 1, 
      rel_heights = c(0.2, 0.8)
    )
    
    right_side <- plot_grid(
      max_like_genos_plot,
      max_post_genos_plot,
      post_freq_plot,
      ncol = 1,
      rel_heights = c(0.3, 0.3, 0.4)
    )
    
     # and, finally, put those all together
    plot_grid(
      left_side,
      right_side,
      nrow = 1,
      rel_widths = c(0.65, 0.35)
    )
    
  },
  height = 750
)
```
